File: backtest-example/src/main/java/com/yuanzhixiang/bt/example/kit/AccountKit.java
package com.yuanzhixiang.bt.example.kit;

import com.yuanzhixiang.bt.engine.BackTestException;

import java.math.RoundingMode;

import static java.math.BigDecimal.valueOf;

/**
 * @author Yuan Zhixiang
 */
public class AccountKit {

    /**
     * Calculate quantity
     *
     * @param price   buy price
     * @param percent percent of position, range in (0, 1]
     * @return calculate buy quantity
     */
    public static double buy(double balance, double price, double percent) {
        if (percent > 1 || percent <= 0) {
            throw new BackTestException("Percentage must be between (0, 1]");
        }

        // Calculate the quantity that can be bought
        int quantity = valueOf(balance)
                .multiply(valueOf(percent))
                .divide(valueOf(price), 0, RoundingMode.DOWN)
                .intValue();
        quantity = (quantity / 100) * 100;

        return quantity;
    }

}

File: backtest-example/src/main/java/com/yuanzhixiang/bt/example/kit/AccountKit.kt
package com.yuanzhixiang.bt.example.kit

import com.yuanzhixiang.bt.engine.BackTestException
import java.math.BigDecimal
import java.math.RoundingMode

/**
 * @author Yuan Zhixiang
 */
object AccountKit {
    /**
     * Calculate quantity
     *
     * @param price   buy price
     * @param percent percent of position, range in (0, 1]
     * @return calculate buy quantity
     */
    @JvmStatic
    fun buy(balance: BigDecimal, price: BigDecimal, percent: BigDecimal): BigDecimal {
        if (percent > BigDecimal.ONE || percent <= BigDecimal.ZERO) {
            throw BackTestException("Percentage must be between (0, 1]")
        }

        // Calculate the quantity that can be bought
        var quantity: Int = balance
            .multiply(percent)
            .divide(price, 0, RoundingMode.DOWN)
            .toInt()
        quantity = quantity / 100 * 100
        return quantity.toBigDecimal()
    }
}